src-control/Control/ActionRegistry.hs

```diff
@@ -0,0 +1,377 @@
+-- in the presence of (a-)synchronous exceptions.
+--
+-- This module is heavily inspired by:
+--
+-- * [resource-registry](https://github.com/IntersectMBO/io-classes-extra/blob/main/resource-registry/src/Control/ResourceRegistry.hs)
+--
+-- * [resourcet](https://hackage.haskell.org/package/resourcet)
+module Control.ActionRegistry (
+    -- * Modify mutable state
+    -- $modify-mutable-state #modify-mutable-state#
+    modifyWithActionRegistry
+  , modifyWithActionRegistry_
+    -- * Action registry
+  , ActionRegistry
+    -- * Runners
+  , withActionRegistry
+  , unsafeNewActionRegistry
+  , unsafeFinaliseActionRegistry
+    -- * Registering actions
+    -- $registering-actions #registering-actions#
+  , withRollback
+  , withRollback_
+  , withRollbackMaybe
+  , withRollbackEither
+  , withRollbackFun
+  , delay
+  ) where
+
+import           Control.Monad.Class.MonadThrow
+import           Control.Monad.Primitive
+import           Data.Either (lefts)
+import           Data.List.NonEmpty as NE
+import           Data.Primitive.MutVar
+
+-- TODO: replace TempRegistry by ActionRegistry
+
+{-------------------------------------------------------------------------------
+  Modify mutable state
+-------------------------------------------------------------------------------}
+
+{- $modify-mutable-state
+
+  When a piece of mutable state holding system resources is updated, then it is
+  important to guarantee in the presence of (a-)synchronous exceptions that:
+
+  1. Allocated resources end up in the state
+  2. Freed resources are removed from the state
+
+  Consider the example program below. We have some mutable @State@ that holds a
+  file handle/descriptor. We want to mutate this state by closing the current
+  handle, and replacing it by a newly opened handle. Using the tools at our
+  disposal in "Control.ActionRegistry", we guarantee (1) and (2).
+
+  @
+    type State = MVar Handle
+
+    example :: State -> IO ()
+    example st =
+      'modifyWithActionRegistry_'
+        (takeMVar st)
+        (putMVar st)
+        $ \\reg h -> do
+          h' <- 'withRollback' reg
+                  (openFile  "file.txt" ReadWriteMode)
+                  hClose
+          'delay' reg (hClose h)
+          pure h'
+  @
+
+  What is also nice about this examples is that it is atomic: other threads will
+  not be able to see the updated @State@ until 'modifyWithActionRegistry_' has
+  exited and the necessary side effects have been performed. Of course, another
+  thread *could* observe that the @file.txt@ was created before
+  'modifyWithActionRegistry_' has exited, but the assumption is that the threads
+  in our program are cooperative. It is up to the user to ensure that actions
+  that are performed as part of the state update do not conflict with other
+  actions.
+-}
+
+-- TODO: add assertions that allocated resources end up in the final state, and
+-- that temporarily freed resources are removed from the final state.
```

Comment:
This is a feature that the consensus `ResourceRegistry` has. Due to time constraints, I think it is best to leave this as a TODO, because it would be nice to have, but not strictly necessary. It's mostly nice for debugging purposes